api/src/main/java/org/eclipse/microprofile/config/Config.java

```diff
@@ -291,4 +303,181 @@
+     * server.ports.0=8081
+     * server.ports.1=8082
+     * </pre>
+     * <p>
+     * {@link Type#VALUE} - the nodes {@code server.port} and {@code server.host} are values
+     * <pre>
+     * server.port=8080
+     * server.host=localhost
+     * </pre>
+     *
+     * @return {@code true} if the node has direct value, {@code false} otherwise.
+     */
+    boolean hasValue();
+
+    /**
+     * Typed value created using a converter function.
+     * The converter is called only if this config node exists.
+     *
+     * @param converter to create an instance from config node
+     * @param <T> type of the object
+     * @return converted value of this node, or an empty optional if this node does not exist
+     * @throws java.lang.IllegalArgumentException if this config node cannot be converted to the desired type
+     */
+    <T> Optional<T> as(Function<Config, T> converter);
+
+    /**
+     * Typed value created using a configured converter.
+     *
+     * @param type class to convert to
+     * @param <T> type of the object
+     * @return converted value of this node, or empty optional if this node does not exist
+     * @throws java.lang.IllegalArgumentException if this config node cannot be converted to the desired type
+     */
+    <T> Optional<T> as(Class<T> type);
+
+    /**
+     * Map to a list of typed values.
+     * This method is only available if the current node is a {@link org.eclipse.microprofile.config.Config.Type#LIST}
+     * or if it has a direct value. In case a direct value of type String exists, it expects comma separated elements.
+     *
+     * @param type class to convert each element to
+     * @param <T> type of the object
+     * @return list of typed values
+     * @throws java.lang.IllegalArgumentException if this config node cannot be converted to the desired type
+     */
+    <T> Optional<List<T>> asList(Class<T> type);
+
+    /**
+     * Contains the (known) config values as a map of key->value pairs.
+     *
+     * @return map of sub keys of this config node, or empty if this node does not exist
+     */
+    Optional<Map<String, String>> asMap();
+
+    /**
+     * A list of child nodes.
+     * In case this node is {@link org.eclipse.microprofile.config.Config.Type#LIST} returns the list in the correct order.
+     * In case this node is {@link org.eclipse.microprofile.config.Config.Type#OBJECT} returns all direct child nodes
+     *      (in an unknown order)
+     *
+     * @return list of child nodes, or empty if this node does not exist
+     */
+    Optional<List<Config>> asNodeList();
+
+    /*
+     * Shortcut helper methods
+     */
+    default Optional<String> asString() {
+        return as(String.class);
+    }
+
+    default Optional<Integer> asInt() {
+        return as(Integer.class);
+    }
+
+    /**
+     * Config node type.
+     */
+    enum Type {
+        /**
+         * Object node with named members and a possible direct value.
+         */
+        OBJECT,
```

Comment:
I like the Type structure, but not the enum value names. I think it is pretty much: Parent, List, Leaf, Empty

Reply:
Parent is not a very good name, as it can have a value (it is basically a tree node that may contain child nodes and a value).
Leaf is OK. Empty vs. Missing - I think missing is more spot on (as this config node is missing from configuration). Empty may mean something different (e.g. the node exists, but is empty).
The values should still be all upper case (best practice for naming enum values)

Reply:
Renamed value to `LEAF`